fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,1114 @@
+
+from math import ceil
+from time import time
+
+from sklearn.utils import check_scalar
+
+from ._constants import (
+    _IMPORT_ERROR_MESSAGE,
+    _KWARG_ERROR_MESSAGE,
+    _PREDICTION_FUNCTION_AMBIGUOUS,
+    _PROGRESS_UPDATE,
+    _CALLBACK_RETURNS_ERROR,
+)
+from ._backend_engine import BackendEngine
+from ._pytorch_engine import PytorchEngine
+from ._tensorflow_engine import TensorflowEngine
+from ._preprocessor import (
+    FloatTransformer,
+    _get_type,
+)
+from sklearn.base import (
+    ClassifierMixin,
+    RegressorMixin,
+    BaseEstimator,
+    TransformerMixin,
+)
+from sklearn.utils.validation import (
+    check_is_fitted,
+    check_random_state,
+    check_array,
+)
+from sklearn.exceptions import NotFittedError
+from numpy import zeros, argmax, arange
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class _AdversarialFairness(BaseEstimator):
+    r"""Train PyTorch or Tensorflow predictors while mitigating unfairness .
+
+    This model implements the supervised learning method proposed in
+    `"Mitigating Unwanted Biases with Adversarial Learning"
+    <https://dl.acm.org/doi/pdf/10.1145/3278721.3278779>`_ [1]_.
+    This algorithm takes as input two neural network
+    models, a predictor model and an adversarial model, defined either as a
+    `PyTorch module
+    <https://pytorch.org/docs/stable/generated/torch.nn.Module.html>`_ or
+    `Tensorflow2 model
+    <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`_. You train this
+    predictor using an API that is similar to estimators in `sklearn`.
+
+    As per [1]_, the neural network models :code:`predictor_model` and
+    :code:`adversary_model` should not have a discrete prediction at the end of the
+    model, as gradients do not propagate through a discretization step.
+    When discrete predictions are required, as is the case in binary and
+    multiclass classification, please just supply the predictor model that
+    produces continuous values, e.g., using a sigmoidal or soft-max
+    activation, and specify the appropriate final transformation separately via
+    parameter :code:`prediction_function` (or let us automatically infer this).
+
+    The distribution types of :code:`y` and :code:`sensitive_features`
+    are set by their preprocessor :code:`y_transform` and :code:`sf_transform`
+    respectively. The default transformer :code:`"auto"`
+    attempts to automatically infer
+    whether to assume binomial, multinomial, or normally distributed data.
+    You can force the transformer to assume one of the above by passing
+    :code:`"binary"`, :code:`"category"`,
+    :code:`"continuous"` instead. Better yet,
+    you can pass :code:`"one_hot_encoder"` or :code:`None` en choose a
+    preprocessing technique.
+    Loss functions and decision functions
+    for the target and sensitive features are also predefined for inferred data
+    types, and are hence automatically inferred.
+
+    Parameters
```

Comment:
Please adjust the parameter descriptions similar to `AdversarialFairnessClassifier`.